akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -3726,4 +3726,64 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
+   * That won't work for the use case of writing large files.
+   * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+   * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+   * This custom flow uses custom aggregator to support such use cases.
+   * Upstream inputs are continuously aggregated as they arrive.
+   * The currently aggregated elements can be terminated and emitted based on custom conditions
+   *
+   * @param allocate    allocate the initial data structure for aggregated elements
+   * @param aggregate   update the aggregated elements, return true if ready to emit after update
+   * @param emitOnTimer decide whether the currently aggregated elements can be emitted on each timer event
+   * @param harvest     this is invoked before emit within the current stage/operator
+   */
+  def aggregateWithBoundary[Agg, Emit](allocate: => Agg)(
+      aggregate: (Agg, Out) => Boolean,
+      harvest: Agg => Emit,
+      emitOnTimer: Option[(Agg => Boolean, FiniteDuration)]): Repr[Emit] =
+    via(AggregateWithBoundary(() => allocate, aggregate, harvest, emitOnTimer))
+
+  /**
+   * This is a convenient wrapper of [[aggregateWithBoundary]] to handle 2 additional time constraints
+   * @param maxGap        the gap allowed between consecutive aggregate operations
+   * @param maxDuration   the duration of the sequence of aggregate operations from initial seed until emit is triggered
+   * @param interval      interval of the timer to check the maxGap and maxDuration condition
+   * @param currentTimeMs source of the system time, can be simulated time in testing
+   */
+  def aggregateWithTimeBoundary[Agg, Emit](allocate: => Agg)(
+      aggregate: (Agg, Out) => Boolean,
+      harvest: Agg => Emit,
+      maxGap: Option[FiniteDuration],
+      maxDuration: Option[FiniteDuration],
+      interval: FiniteDuration,
+      currentTimeMs: => Long): Repr[Emit] = {
+    require(
+      maxDuration.nonEmpty || maxGap.nonEmpty,
+      s"required timing condition maxGap and maxDuration are both missing, use aggregateWithBoundary if it's intended")
+
+    class ValueTimeWrapper[T](var value: T) {
+      var firstTime: Long = -1
+      var lastTime: Long = -1
+      def updateTime(time: Long): Unit = {
+        if (firstTime == -1) firstTime = time
+        lastTime = time
+      }
+    }
+
+    aggregateWithBoundary(allocate = new ValueTimeWrapper(value = allocate))(aggregate = (agg, in) => {
+      agg.updateTime(currentTimeMs)
+      // user provided Agg type must be mutable
+      aggregate(agg.value, in)
+    }, harvest = agg => harvest(agg.value), emitOnTimer = Some((agg => {
+      val currentTime = currentTimeMs
+      maxDuration.exists(md => currentTime - agg.firstTime >= md.toMillis) ||
+      maxGap.exists(mg => currentTime - agg.lastTime >= mg.toMillis)
+    }, interval)))
+  }
+
```

Comment:
Once we land on exact method params and docs etc there should be delegate factories for these but with suitable types for usage from Java (`Function`, `Optional`, `java.time.Duration`) in `akka.stream.javadsl.Flow` and `akka.stream.javadsl.Source` as well.

Reply:
do we need to add java tests or just passing compile is sufficient?

Reply:
Need some help and examples to add the javadsl interface seems for this.
1. zero parameter function `allocate: () => Agg`, found some example using `function.Creator[Agg]`, is this the right one?
2.`emitOnTimer: Option[(Agg => Boolean, FiniteDuration)]`, what's the right interface for javadsl?

Shall we address javadsl as another PR later?

Reply:
Java API needs to go in this PR, I think just compiling will be fine since there will be a minimal amount of logic in there, just type conversions.

`() => Agg` should be `java.util.function.Supplier<Agg>`
`Option[(Agg => Boolean, FiniteDuration)]` could be `Optional<Pair<java.util.function.Predicate<Agg>, java.time.Duration>>` but since that is a bit messy for users to actually construct in Java I think it would be better with one overload without timeout and one with separate `Duration` and `Predicate` parameters and hiding the optional/option part behind that.

Reply:
Added javadsl methods. I just added one version with the same doc in both `Flow` and `Source`. User can always have other wrappers to customize however they want to overload. Otherwise, copying two methods with small variants in 4 places seems not worth the maintenance overhead.

Reply:
And for the same reason, I appreciate the decision to not include the method with 2 timeouts :)

Reply:
Fair enough, it is a power user operator anyway, and we can always add convenient factory at any time in the future.